[RFC] bcm63xx: add NAND support (Comtrend VR-3032u)

Noltari opened the pull request:
This PR adds NAND support for bcm63xx/smp.

It took me some time to figure out a way to get separate JFFS2 partitions for CFE RAM and kernel.
Having separate JFFS2 partitions is great, since we only need to flash CFE RAM the first time that we install OpenWrt from the bootloader via http.
For sysupgrading, CFE RAM won't be reflashed.

Comtrend VR-3032u flashing and bootlog:
```
HELO
CPUI
L1CI
HELO
CPUI
L1CI
DRAM
----
PHYS
STRF
400H
PHYE
DDR2
SIZ4
SIZ3
SIZ2
SIZ1
DINT
USYN
LSYN
MFAS
LMBE
RACE
PASS
----
ZBSS
CODE
DATA
L12F
MAIN
FPS0
BT26
0001
STOP
NAN9
NAN3
RFS1
NAN5


CFE version 1.0.38-112.118-11 for BCM963268 (32bit,SP,BE)
Build Date: �12月 27 11:54:52 CST 2012 (camille@SW1-BCM-02)
Copyright (C) 2000-2011 Broadcom Corporation.

NAND flash device: name , id 0xc2f1 block 128KB size 131072KB
Chip ID: BCM63168D0, MIPS: 400MHz, DDR: 400MHz, Bus: 200MHz
Main Thread: TP0
Memory Test Passed
Total Memory: 67108864 bytes (64MB)
Boot Address: 0xb8000000

Board IP address                  : 192.168.1.4:ffffff00  
Host IP address                   : 192.168.1.8  
Gateway IP address                : 192.168.1.1  
Run from flash/host (f/h)         : f  
Default host run file name        : vr3032u.elf  
Default host flash file name      : bcm963xx_fs_kernel  
Boot delay (0-9 seconds)          : 1  
Board Id (0-28)                   : 963168M-1841N1  
Number of MAC Addresses (1-32)    : 32  
Base MAC Address                  : f8:8e:85:4e:7f:0d  
PSI Size (1-64) KBytes            : 24  
Enable Backup PSI [0|1]           : 1  
System Log Size (0-256) KBytes    : 0  
Main Thread Number [0|1]          : 0  
WLAN Password                     : 4Tesh8bwme6ScTrhnxPv  
Serial Number                     : 14A3032UXXF-AA010739  

*** Press any key to stop auto run (1 seconds) ***
Auto run second count down: 1
 Port 1 link UP 
11
CFE> web info: Waiting for connection on socket 0.
CFE>    web info: Waiting for connection on socket 1.
CFE>    web info: Waiting for connection on socket 0.
CFE>    web info: Waiting for connection on socket 2.
CFE>    web info: Upload 8650772 bytes, flash image format.
CFE>    
Flashing root file system at 0xb8020000: ..................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

.


Old image, deleting data and secondary file system partitions
nandflash_block_erase(): Attempt to erase bad nand block 616

Error erasing flash block, blk=616

Resetting board...HELO
CPUI
L1CI
HELO
CPUI
L1CI
DRAM
----
PHYS
STRF
400H
PHYE
DDR2
SIZ4
SIZ3
SIZ2
SIZ1
DINT
USYN
LSYN
MFAS
LMBE
RACE
PASS
----
ZBSS
CODE
DATA
L12F
MAIN
FPS0
BT27
0001
STOP
NAN9
NAN3
RFS1
NAN5


CFE version 1.0.38-112.118-11 for BCM963268 (32bit,SP,BE)
Build Date: �12月 27 11:54:52 CST 2012 (camille@SW1-BCM-02)
Copyright (C) 2000-2011 Broadcom Corporation.

NAND flash device: name , id 0xc2f1 block 128KB size 131072KB
Chip ID: BCM63168D0, MIPS: 400MHz, DDR: 400MHz, Bus: 200MHz
Main Thread: TP0
Memory Test Passed
Total Memory: 67108864 bytes (64MB)
Boot Address: 0xb8000000

Board IP address                  : 192.168.1.4:ffffff00  
Host IP address                   : 192.168.1.8  
Gateway IP address                : 192.168.1.1  
Run from flash/host (f/h)         : f  
Default host run file name        : vr3032u.elf  
Default host flash file name      : bcm963xx_fs_kernel  
Boot delay (0-9 seconds)          : 1  
Board Id (0-28)                   : 963168M-1841N1  
Number of MAC Addresses (1-32)    : 32  
Base MAC Address                  : f8:8e:85:4e:7f:0d  
PSI Size (1-64) KBytes            : 24  
Enable Backup PSI [0|1]           : 1  
System Log Size (0-256) KBytes    : 0  
Main Thread Number [0|1]          : 0  
WLAN Password                     : 4Tesh8bwme6ScTrhnxPv  
Serial Number                     : 14A3032UXXF-AA010739  

*** Press any key to stop auto run (1 seconds) ***
Auto run second count down: 110
Booting from only image (0xb8020000) ...
Decompression OK!
Entry at 0x80a00000
Closing network.
Disabling Switch ports.
Flushing Receive Buffers...
0 buffers found.
Closing DMA Channels.
Starting program at 0x80a00000
[    0.000000] Linux version 5.4.36 (noltari@skynet) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r12787-af5ccfbac7)) #0 SMP Sun May 3 15:50:26 2020
[    0.000000] Detected Broadcom 0x63268 CPU revision d0
[    0.000000] CPU frequency is 400 MHz
[    0.000000] 64MB of RAM installed
[    0.000000] board_bcm963xx: Boot address 0xb8000000
[    0.000000] board_bcm963xx: CFE version: 1.0.38-112.118-11
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0002a080 (Broadcom BMIPS4350)
[    0.000000] board: board name: 963168M-1841N1
[    0.000000] MIPS: machine is Comtrend VR-3032u
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 16 bytes.
[    0.000000] Primary data cache 32kB, 2-way, VIPT, cache aliases, linesize 16 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] percpu: Embedded 14 pages/cpu s26048 r8192 d23104 u57344
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16240
[    0.000000] Kernel command line: rootfstype=squashfs,ubifs noinitrd console=ttyS0,115200
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 56108K/65536K available (5689K kernel code, 212K rwdata, 1216K rodata, 1324K init, 206K bss, 9428K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=16, Order=0-3, MinObjects=0, CPUs=2, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 25 jiffies.
[    0.000000] NR_IRQS: 256
[    0.000000] genirq: irq_chip bcm6345-periph-intc did not update eff. affinity mask of irq 52
[    0.000000] random: get_random_bytes called from start_kernel+0x338/0x568 with crng_init=0
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 9556302233 ns
[    0.000013] sched_clock: 32 bits at 200MHz, resolution 5ns, wraps every 10737418237ns
[    0.008134] Calibrating delay loop... 397.82 BogoMIPS (lpj=795648)
[    0.046441] pid_max: default: 32768 minimum: 301
[    0.051520] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.059059] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.071876] rcu: Hierarchical SRCU implementation.
[    0.077972] smp: Bringing up secondary CPUs ...
[    0.084152] SMP: Booting CPU1...
[   10.492838] Primary instruction cache 64kB, VIPT, 4-way, linesize 16 bytes.
[   10.492856] Primary data cache 32kB, 2-way, VIPT, cache aliases, linesize 16 bytes
[   10.493129] CPU1 revision is: 0002a080 (Broadcom BMIPS4350)
[    0.119645] Synchronize counters for CPU 1: 
[    0.119664] SMP: CPU1 is running
[    0.119685] done.
[    0.150193] smp: Brought up 1 node, 2 CPUs
[    0.165321] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.175380] futex hash table entries: 512 (order: 1, 8192 bytes, linear)
[    0.182548] pinctrl core: initialized pinctrl subsystem
[    0.188923] NET: Registered protocol family 16
[    0.271662] clocksource: Switched to clocksource MIPS
[    0.279785] NET: Registered protocol family 2
[    0.285594] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 6144 bytes, linear)
[    0.294307] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.302224] TCP bind hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.309533] TCP: Hash tables configured (established 1024 bind 1024)
[    0.316504] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.323322] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.331016] NET: Registered protocol family 1
[    0.335554] PCI: CLS 0 bytes, default 16
[    0.345652] workingset: timestamp_bits=14 max_order=14 bucket_order=0
[    0.365688] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.371705] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.402434] random: fast init done
[    0.409611] bcm63268-pinctrl 100000c0.pin-controller: registered at mmio (ptrval)
[    0.418677] 10000180.serial: ttyS0 at MMIO 0x10000180 (irq = 13, base_baud = 1562500) is a bcm63xx_uart
[    0.428426] printk: console [ttyS0] enabled
[    0.428426] printk: console [ttyS0] enabled
[    0.436993] printk: bootconsole [early0] disabled
[    0.436993] printk: bootconsole [early0] disabled
[    0.451459] nand: device found, Manufacturer ID: 0xc2, Chip ID: 0xf1
[    0.458015] nand: Macronix NAND 128MiB 3,3V 8-bit
[    0.462835] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
[    0.470652] bcm6368_nand 10000200.nand: detected 128MiB total, 128KiB blocks, 2KiB pages, 16B OOB, 8-bit, Hamming ECC
[    0.482586] Bad block table found at page 65472, version 0x01
[    0.489193] Bad block table found at page 65408, version 0x01
[    0.495927] nand_read_bbt: bad block at 0x000004d00000
[    0.502106] 2 fixed-partitions partitions found on MTD device brcmnand.0
[    0.509053] Creating 2 MTD partitions on "brcmnand.0":
[    0.514333] 0x000000000000-0x000000020000 : "cferom"
[    0.521164] 0x000000020000-0x000007f00000 : "wfi"
[    0.734481] 4 bcm-wfi-fw partitions found on MTD device wfi
[    0.740198] Creating 4 MTD partitions on "wfi":
[    0.744866] 0x000000000000-0x000000160000 : "cferam"
[    0.751860] 0x000000160000-0x000007ee0000 : "firmware"
[    0.759787] 0x000000160000-0x000000660000 : "kernel"
[    0.766589] 0x000000660000-0x000007ee0000 : "ubi"
[    0.782976] bcm63xx-spi 10000800.spi: at [mem 0x10000800-0x10000f0b flags 0x200] (irq 88, FIFOs size 542)
[    0.796742] libphy: Fixed MDIO Bus: probed
[    0.844154] b53_common: found switch: BCM63xx, rev 0
[    0.849739] bcm63xx-wdt bcm63xx-wdt:  started, timer margin: 30 sec
[    0.858011] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    0.865203] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    0.872339] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    0.879511] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    0.886673] leds-gpio leds-gpio.0: Skipping unavailable LED gpio 0 ((null))
[    0.896047] NET: Registered protocol family 10
[    0.912055] Segment Routing with IPv6
[    0.916002] NET: Registered protocol family 17
[    0.920632] 8021q: 802.1Q VLAN Support v1.8
[    0.946387] UBI: auto-attach mtd5
[    0.949865] ubi0: attaching mtd5
[    0.977120] UBI: EOF marker found, PEBs from 18 will be erased
[    0.985074] ubi0: scanning is finished
[    1.030291] ubi0: volume 1 ("rootfs_data") re-sized from 9 to 924 LEBs
[    1.039060] ubi0: attached mtd5 (name "ubi", size 120 MiB)
[    1.044736] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    1.051779] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    1.058759] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    1.065924] ubi0: good PEBs: 963, bad PEBs: 1, corrupted PEBs: 0
[    1.072107] ubi0: user volume: 2, internal volumes: 1, max. volumes count: 128
[    1.079547] ubi0: max/mean erase counter: 1/0, WL threshold: 4096, image sequence number: 1588521026
[    1.088954] ubi0: available PEBs: 0, total reserved PEBs: 963, PEBs reserved for bad PEB handling: 19
[    1.098501] ubi0: background thread "ubi_bgt0d" started, PID 400
[    1.102414] block ubiblock0_0: created from ubi0:0(rootfs)
[    1.110415] ubiblock: device ubiblock0_0 (rootfs) set to be root filesystem
[    1.134435] VFS: Mounted root (squashfs filesystem) readonly on device 254:0.
[    1.160924] Freeing unused kernel memory: 1324K
[    1.165612] This architecture does not have kernel memory protection.
[    1.172240] Run /sbin/init as init process
[    2.264190] random: crng init done
[    2.391965] init: Console is alive
[    2.396036] init: - watchdog -
[    3.135304] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.267069] usbcore: registered new interface driver usbfs
[    3.272905] usbcore: registered new interface driver hub
[    3.278555] usbcore: registered new device driver usb
[    3.295111] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    3.304011] ehci-fsl: Freescale EHCI Host controller driver
[    3.312038] ehci-platform: EHCI generic platform driver
[    3.423402] ehci-platform ehci-platform: EHCI Host Controller
[    3.429403] ehci-platform ehci-platform: new USB bus registered, assigned bus number 1
[    3.437910] ehci-platform ehci-platform: irq 18, io mem 0xb0002500
[    3.459377] ehci-platform ehci-platform: USB 2.0 started, EHCI 1.00, overcurrent ignored
[    3.469022] hub 1-0:1.0: USB hub found
[    3.472999] hub 1-0:1.0: 1 port detected
[    3.486628] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    3.494788] ohci-platform: OHCI generic platform driver
[    3.500423] ohci-platform ohci-platform: Generic Platform OHCI controller
[    3.507479] ohci-platform ohci-platform: new USB bus registered, assigned bus number 2
[    3.515795] ohci-platform ohci-platform: irq 17, io mem 0xb0002600
[    3.584865] hub 2-0:1.0: USB hub found
[    3.588841] hub 2-0:1.0: 1 port detected
[    3.594471] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    3.613183] init: - preinit -
[    5.773439] IPv6: ADDRCONF(NETDEV_CHANGE): eth0.1: link becomes ready
[    5.780470] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[   10.302169] UBIFS (ubi0:1): default file-system created
[   10.308986] UBIFS (ubi0:1): Mounting in unauthenticated mode
[   10.315245] UBIFS (ubi0:1): background thread "ubifs_bgt0_1" started, PID 514
[   10.454837] UBIFS (ubi0:1): UBIFS: mounted UBI device 0, volume 1, name "rootfs_data"
[   10.462903] UBIFS (ubi0:1): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
[   10.473093] UBIFS (ubi0:1): FS size: 115929088 bytes (110 MiB, 913 LEBs), journal size 5840896 bytes (5 MiB, 46 LEBs)
[   10.484012] UBIFS (ubi0:1): reserved for root: 4952683 bytes (4836 KiB)
[   10.490839] UBIFS (ubi0:1): media format: w4/r0 (latest is w5/r0), UUID 1099DFBA-AB6D-4C1E-8623-BD837039DC12, small LPT model
[   10.505788] mount_root: overlay filesystem has not been fully initialized yet
[   10.513900] mount_root: switching to ubifs overlay
[   10.551139] overlayfs: upper fs does not support xattr, falling back to index=off and metacopy=off.
[   10.565387] urandom-seed: Seed file not found (/etc/urandom.seed)
[   10.770285] procd: - early -
[   10.773397] procd: - watchdog -
[   11.437799] procd: - watchdog -
[   11.441450] procd: - ubus -
[   11.651586] procd: - init -
Please press Enter to activate this console.
[   12.623110] kmodloader: loading kernel modules from /etc/modules.d/*
[   12.821939] xt_time: kernel timezone is -0000
[   12.869367] PPP generic driver version 2.4.2
[   12.877629] NET: Registered protocol family 24
[   12.908551] kmodloader: done loading kernel modules from /etc/modules.d/*
[   12.937813] urngd: v1.0.2 started.
[   80.220444] br-lan: port 1(eth0.1) entered blocking state
[   80.226245] br-lan: port 1(eth0.1) entered disabled state
[   80.232690] device eth0.1 entered promiscuous mode
[   80.237641] device eth0 entered promiscuous mode
[   80.273633] br-lan: port 1(eth0.1) entered blocking state
[   80.279219] br-lan: port 1(eth0.1) entered forwarding state
[   81.228876] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
```

adrianschmutzler commented on target/linux/bcm63xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -1,9 +1,19 @@
 	case "$(get_magic_word "$1")" in
 		3600|3700|3800)
-			# 6348GW-11 boards use openwrt-96348GW-11-squashfs-cfe.bin files
+			# CFE tag versions
```
OT: Do you have any idea how many boards are covered here? If it's not too painful, I'd consider switching to board_name for them as well, so we end up with only one case?

Noltari replied:
Basically every bcm63xx board is covered here, since all of them rely on the CFE Tag.

adrianschmutzler commented on target/linux/bcm63xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -13,4 +23,50 @@ platform_check_image() {
+	local tar_file="$1"
+	local kernel_mtd="$(find_mtd_index $CI_KERNPART)"
+
+	if [ ! "$kernel_mtd" ]; then
```
-z ?

Noltari replied:
I will change it.
FYI, this was copied from https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/nand.sh#L45

adrianschmutzler replied:
Hmm, looks like this should be patched.

adrianschmutzler commented on target/linux/bcm63xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -13,4 +23,50 @@ platform_check_image() {
+	board_dir=${board_dir%/}
+
+	local kernel_length=`(tar xf $tar_file ${board_dir}/kernel -O | wc -c) 2> /dev/null`
+	local rootfs_length=`(tar xf $tar_file ${board_dir}/root -O | wc -c) 2> /dev/null`
```
Consider replacing the backticks by $() syntax:

```
	local kernel_length=$((tar xf $tar_file ${board_dir}/kernel -O | wc -c) 2> /dev/null)
	local rootfs_length=$((tar xf $tar_file ${board_dir}/root -O | wc -c) 2> /dev/null)
```

We've done that for other files maybe half a year ago IIRC.

Noltari replied:
Ok, I change it.
FYI, this was copied from: https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/nand.sh#L252

adrianschmutzler replied:
Thanks, will send a patch as well. I think back then the changes were limited to target directory.

Noltari replied:
After testing your suggestion it doesn't work, so I'm reverting this change.

adrianschmutzler replied:
Sorry, it's kind of a deja vu:
https://github.com/openwrt/openwrt/commit/291d79935e9c4a974b037b6bed733ce92d4242e3#diff-a6a780f4082e88756e922a1f3ec5e2ab

"$((" should be replaced by "$( (", however, in the references PR it was decided to drop the subshell entirely. Sorry for the wrong hint, now I will go and fix my patches just sent :-(

adrianschmutzler commented on target/linux/bcm63xx/dts/bcm63268-comtrend-vr-3032u.dts:
```diff
@@ -0,0 +1,176 @@
+/dts-v1/;
```
License?

Noltari replied:
Yes, I missed that :D

adrianschmutzler commented on target/linux/bcm63xx/dts/bcm63268-comtrend-vr-3032u.dts:
```diff
@@ -0,0 +1,176 @@
+};
+
+&leds {
+	status = "ok";
```
ok -> okay

Noltari replied:
Ok, but this needs to be fixed on most of bcm63xx dts files...

adrianschmutzler replied:
I don't think it breaks something yet, but for newly added devices I try to have it "right" ...

adrianschmutzler commented on target/linux/bcm63xx/dts/bcm63268-comtrend-vr-3032u.dts:
```diff
@@ -0,0 +1,176 @@
+	led@2 {
+		reg = <2>;
+		active-low;
+		label = "VR-3032u:red:inet";
```
if model name was in lower-case here, one could derive it from the device name in base-files scripts, so I typically recommend that ...

Noltari replied:
This happens on nearly every bcm63xx device :P

adrianschmutzler replied:
Indeed, but changing the old ones will break compatibility, while having the new ones matching the model part doesn't have a downside. I'm just looking at being able to use this at least for new devices: https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/generic/base-files/etc/board.d/01_leds#L8

adrianschmutzler commented on target/linux/bcm63xx/dts/bcm63268-comtrend-vr-3032u.dts:
```diff
@@ -0,0 +1,176 @@
+};
+
+&uart0 {
+	status = "ok";
```
ok -> okay

Noltari replied:
Ok, but this needs to be fixed on most of bcm63xx dts files...

adrianschmutzler commented on target/linux/bcm63xx/dts/bcm63268-comtrend-vr-3032u.dts:
```diff
@@ -0,0 +1,176 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cferom@0 {
```
partition@0 ?

Noltari replied:
Ok, I will change that

adrianschmutzler commented on target/linux/bcm63xx/dts/bcm63268-comtrend-vr-3032u.dts:
```diff
@@ -0,0 +1,176 @@
+			wfi@20000 {
+				compatible = "brcm,wfi";
+				label = "wfi";
+				reg = <0x0020000 0x7ee0000>;
```
Is it really only 0x7f00000 or is this a typo?

Noltari replied:
Yes, this is on purpose, since the BBT is located at the end of the flash.

adrianschmutzler commented on target/linux/bcm63xx/base-files/etc/diag.sh:
```diff
@@ -61,6 +61,9 @@ set_state() {
 		status_led="VR-3026e:green:power"
 		;;
+	comtrend,vr-3032u)
+		status_led="VR-3032u:green:power"
```
OT: This file looks like it would be relatively trivial to use generic diag.sh on this target. Do you think it would be worth it, or are you planning to replace with bmips soon anyway?

Noltari replied:
I would like to switch to bmips target, but @KanjiMonster should also agree with this decission, and so far I haven't been able to contact him about it...

Noltari commented:
@adrianschmutzler I updated PR with all your comments and also added support for older NAND controllers.

Noltari committed:
bcm63xx: add support for Comtrend VR-3032u

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
bcm63xx-cfe: add package with CFE RAM binaries

CFE RAM is a second stage bootloader which is usually loaded by CFE ROM
(first stage bootloader) from a JFFS2 partition stored on the NAND.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
scripts: support tags for CFE binaries

CFE loads binaries based on a 12 byte header which corresponds to:
- u32: Load Address
- u32: Entry Address
- u32: Size

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
base-files: switch_to_ramfs: add nand-utils

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
scripts: support CFE WFI images

CFE WFI (Whole Flash Image) consists in 20 byte footer:
- u32: CRC32
- u32: WFI Version
- u32: Chip ID
- u32: Flash Type
- u32: Flags

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
kernel: mtdsplit: add BCM WFI support

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
bcm63xx: smp: add NAND support

NAND controller is present on BCM6328, BCM6362, BCM6368 and BCM63268.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
Merged (367a30389d298948a3c4867dd0261bb942371147)
